wcfsetup/install/files/lib/data/user/User.class.php

```diff
@@ -142,35 +144,39 @@ public function __construct($id, $row = null, DatabaseObject $object = null) {
 	 */
 	public function checkPassword($password) {
 		$isValid = false;
-		$rebuild = false;
 		
-		// check if password is a valid bcrypt hash
-		if (PasswordUtil::isBlowfish($this->password)) {
-			if (PasswordUtil::isDifferentBlowfish($this->password)) {
-				$rebuild = true;
-			}
-			
-			// password is correct
-			if (\hash_equals($this->password, PasswordUtil::getDoubleSaltedHash($password, $this->password))) {
-				$isValid = true;
-			}
+		$manager = PasswordAlgorithmManager::getInstance();
+		
+		// Compatibility for WoltLab Suite < 5.4.
+		if (DoubleBcrypt::isLegacyDoubleBcrypt($this->password)) {
+			$algorithmName = 'DoubleBcrypt';
+			$hash = $this->password;
 		}
 		else {
-			// different encryption type
-			if (PasswordUtil::checkPassword($this->username, $password, $this->password)) {
-				$isValid = true;
-				$rebuild = true;
-			}
+			[$algorithmName, $hash] = \explode(':', $this->password);
```

Comment:
This is not working, if there is a salt after the hash (e.g. `algo:hash:salt`)

Reply:
Good call. Can you try the following, please:
```suggestion
			[$algorithmName, $hash] = \explode(':', $this->password, 2);
```
?

Reply:
This works.